Add bank holidays management in TransXChange

woshilapin opened the pull request:
I'm opening this PR as a draft PR because even if the code might be working fine (at least, the few tests I have seems to work), I'm not satisfied about the structure of the code. I'm open to ideas.

Note: I'm currently working on adding integration tests too.

datanel commented on src/transxchange/bank_holidays.rs:
```diff
@@ -0,0 +1,201 @@
+        .map_err(|e| Error::from_boxed_compat(Box::new(e)))?;
+    let mut day_per_bank_holiday: HashMap<BankHoliday, Vec<Date>> = HashMap::new();
+    for event in region.events {
+        if let Some(dates) = day_per_bank_holiday.get_mut(&event.title) {
```
```rust
day_per_bank_holiday.entry(event.title).or_insert_with(|| vec![]).push(event.date);
```

woshilapin replied:
Nicely done. Thank you.

datanel commented on src/transxchange/bank_holidays.rs:
```diff
@@ -0,0 +1,201 @@
+    Ok(day_per_bank_holiday)
+}
+
+// Generate a list of all fixed dates between to dates.
```
between **two** dates

woshilapin replied:
I'll fix that.

datanel commented on src/transxchange/read.rs:
```diff
@@ -378,90 +378,241 @@ impl Iterator for NaiveDateRange {
+        bank_holidays.insert(BankHoliday::Christmas, christmas_days);
+        let boxing_days = bank_holidays::get_fixed_days(26, 12, validity_period);
+        bank_holidays.insert(BankHoliday::BoxingDay, boxing_days);
+        let days_included =
```
```rust
let days_included = bank_holiday_operation
            .try_only_child("DaysOfOperation")
            .map(|operation| get_bank_holidays_dates(operation, &bank_holidays))
            .unwrap_or_default();
```

woshilapin replied:
Yes, thank you, I tried to apply that everywhere I saw it.

datanel commented on src/transxchange/read.rs:
```diff
@@ -378,90 +378,241 @@ impl Iterator for NaiveDateRange {
+            } else {
+                Default::default()
+            };
+        let days_excluded =
```
same here

woshilapin replied:
See [coment](https://github.com/CanalTP/transit_model/pull/353#discussion_r323260929).

datanel commented on src/transxchange/read.rs:
```diff
@@ -378,90 +378,241 @@ impl Iterator for NaiveDateRange {
-            // TODO: Handle exceptions (see SpecialDaysOperation)
-            // TODO: Handle bank holidays (see BankHolidaysOperation)
+        let is_regular_day = regular_days.contains(&date.weekday());
+        let is_bank_holiday = if let Some((ref included_bank_holidays, _)) = bank_holidays {
```
```rust
let is_bank_holiday = bank_holidays
            .as_ref()
            .map(|(ref included_bank_holidays, _)| included_bank_holidays.contains(&date))
            .unwrap_or_else(|_| false);
        let is_excluded_bank_holiday = bank_holidays
            .as_ref()
            .map(|(_, ref excluded_bank_holidays)| excluded_bank_holidays.contains(&date))
            .unwrap_or_else(|_| false);
```

woshilapin replied:
I really refactored that part so it doesn't really apply anymore but thanks for the alternative. These are the case where I'm not sure which one is the most readable though.

datanel commented on src/transxchange/read.rs:
```diff
@@ -378,90 +378,241 @@ impl Iterator for NaiveDateRange {
+        bank_holidays: &HashMap<BankHoliday, Vec<Date>>,
+        validity_period: &ValidityPeriod,
+    ) -> Result<(HashSet<Date>, HashSet<Date>)> {
+        fn get_bank_holidays_dates(
```
fn inception :scream: 
can you move the nested functions out of `generate_calendar_dates` function? it's difficult to read.
At least `get_regular_days` and `get_bank_holidays`

woshilapin replied:
OK, so I'm working on a big refactor of this all PR actually because I was not satisfied about that too. The idea will be to have a struct to save the information from `OperatingProfile` into a usable structure and only after, propose an iterator inside the validity period (probably replacing the `NaiveDateRange` with something else).

woshilapin commented on src/transxchange/bank_holidays.rs:
```diff
@@ -0,0 +1,201 @@
+) -> Result<HashMap<BankHoliday, Vec<Date>>> {
+    let bank_holidays_file = File::open(bank_holiday_path)?;
+    let region: BankHolidayRegion = serde_json::from_reader(bank_holidays_file)
+        .map_err(|e| Error::from_boxed_compat(Box::new(e)))?;
```
Indeed, I thought I had a problem and this line was suppose to fix it. But id doesn't. It seems I still don't completely understand `failure`.

woshilapin commented on src/transxchange/bank_holidays.rs:
```diff
@@ -0,0 +1,201 @@
+    BattleOfTheBoyne,
+    Summer,
+    SaintAndrews,
+    // Bank Holiday for Saint Andrews, not necessarily on the 30st of November
```
Yes it is.

woshilapin commented:
I did a big refactor of this whole PR by moving the parsing of `OperatingProfile` into it's own module and providing an iterator on it. I addressed most of your comment on the code that was still here. I'm moving the PR from draft to a ready-to-review.
